src: ignore GCC -Wcast-function-type for v8.h

danbev opened the pull request:
Currently, the following warnings are emitted during compilation:
```console
In file included from ../src/string_bytes.h:29,
                 from ../src/string_bytes.cc:22:
../deps/v8/include/v8.h:
In instantiation of ‘void v8::PersistentBase<T>::SetWeak(
    P*,typename v8::WeakCallbackInfo<P>::Callback, v8::WeakCallbackType)
[with P = node::BaseObject;
T = v8::Object;
typename v8::WeakCallbackInfo<P>::Callback =
    void (*)(const v8::WeakCallbackInfo<node::BaseObject>&)]’:
../src/base_object-inl.h:135:42:   required from here
../deps/v8/include/v8.h:10732:16: warning:
cast between incompatible function types from
‘v8::WeakCallbackInfo<node::BaseObject>::Callback’ {aka
‘void (*)(const v8::WeakCallbackInfo<node::BaseObject>&)’} to
‘Callback’ {aka ‘void (*)(const v8::WeakCallbackInfo<void>&)’}
[-Wcast-function-type]
10732 |                reinterpret_cast<Callback>(callback), type);
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
````
And the same from test/cctest/test_aliased_buffer.cc

We have an open pull request against v8 for this:
https://chromium-review.googlesource.com/c/v8/v8/+/2080361

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

bnoordhuis commented:
How does that work? We include v8.h in numerous places, not just those two files. It also wouldn't silence the warnings when building V8 itself, right?

Apropos the V8 CL, you picked Yang as a reviewer but I don't think he is on the V8 team anymore.

danbev commented:
> How does that work? We include v8.h in numerous places, not just those two files.

We have previously added these kinda of ignores in 3ff2aeceba88928f1aa33fe3ff1cc9bf84da739b. At that point in time I only focused on the ones that were being generated. Hopefully this is only temporary and this will be added in upstream v8.

>It also wouldn't silence the warnings when building V8 itself, right?

As far as I can tell `-Wcast-function-type` is not specified when building V8, so these warnings are not generated for V8 itself. For example, looking at the generated make file `out/tools/v8_gypfiles/v8_base_without_compiler.target.mk` it has the following CFLAGS defined:
```console
CFLAGS_Release := \                                                             
        -pthread \                                                              
        -Wno-unused-parameter \                                                 
        -m64 \                                                                  
        -Wno-return-type \                                                      
        -fno-strict-aliasing \                                                  
        -m64 \                                                                  
        -O3 \                                                                   
        -fno-omit-frame-pointer \                                               
        -fdata-sections \                                                       
        -ffunction-sections \                                                   
         \                                                                      
        -O3 
```
Adding `-Wcast-function-type` allows these warning to be generated when building V8.

>Apropos the V8 CL, you picked Yang as a reviewer but I don't think he is on the V8 team anymore.

Thanks, I've updated it now and added someone else.

bnoordhuis commented:
> `-Wcast-function-type` is not specified when building V8

Oh, I understand now - it's stripped from cflags (`-Wextra` implies `-Wcast-function-type`):

https://github.com/nodejs/node/blob/6ed912dd0319475d2c4ebd9991af52420de17fe8/tools/v8_gypfiles/toolchain.gypi#L171

I'm good with the pragma approach if it's temporary.

danbev commented:
Re-run of failing [node-test-commit-linux](https://ci.nodejs.org/job/node-test-commit-linux/34129/) :heavy_check_mark:

danbev commented:
Landed in f21cf456bf36bd73d8cb6715c745e23efc22a29f.
